### PR TITLE
Enclose Logical Expressions with Parentheses

### LINKED ISSUE #232: Bug due to Parentheses Simplification
I have a very simple rule like this:

```
path /profile/{key1} {
  create() {
    isUser(key1);
  }

  update() {
    isUser(key1);
  }
}
```

Using these functions:

```
function isAuth() {
  (auth != null);
}

function isUser(user) {
  isAuth() && (user == auth.uid);
}
```

I'd expect the `.write` output of this to be something along the lines of:

```
(data.val() == null && (auth != null && $key1 == auth.uid))
||
(data.val() != null && newData.val() != null && (auth != null && $key1 == auth.uid))
```

Instead, what I get is (note the missing parentheses):

```
data.val() == null && (auth != null && $key1 == auth.uid)
||
data.val() != null && newData.val() != null && (auth != null && $key1 == auth.uid)
```

Which is not logically the same - not sure how to fix it since `create` and `update` are handled by Bolt itself.

---

Similarly, if I have a compound rule like this:

```
path /users/{key1} {
  read() {
    isUser(key1) || isAdmin();
  }
}
```

Relying on these functions:

```
function isAuth() {
  (auth != null);
}

function isUser(user) {
  isAuth() && (user == auth.uid);
}

function isAdmin() {
  isAuth() && (root.child('admins').child(auth.uid).val() != null);
}
```

The output I get on `.read` is the following:

```
auth != null && $key1 == auth.uid
||
auth != null && root.child('admins').child(auth.uid).val() != null
```

When I would expect it to be:

```
(auth != null && $key1 == auth.uid)
||
(auth != null && root.child('admins').child(auth.uid).val() != null)
```

Adding a redundant boolean check to the function calls:

```
path /users/{key1} {
  read() {
    (isUser(key1) == true) || (isAdmin() == true);
  }
}
```

Produces a logically valid output of:

```
(auth != null && $key1 == auth.uid) == true
||
(auth != null && root.child('access-control').child('admins').child(auth.uid).val() != null) == true
```

However, it's kind of an obscure behavior.

---

Are these bugs known in Bolt? And if so, is there any way around them without patching the core?

### PR DESCRIPTION
**Pull request by alixaxel**
This PR fixes #232 and #230.

For cases where an expression operator is `&&` or `||`, it surrounds the whole expression with `(...)`, respecting left association. The top-most parentheses is unnecessary but I couldn't figure out how to skip it.

### CODE REVIEW

File: src/ast.ts [hunk 1]
```diff
@@ -705,6 +705,11 @@ export function decodeExpression(exp: Exp, outerPrecedence?: number): string {
         decodeExpression(expOp.args[1], innerPrecedence) + ' : ' +
         decodeExpression(expOp.args[2], innerPrecedence);
     }
+
+    if ((innerPrecedence >= outerPrecedence) && ((expOp.op === '&&') || (expOp.op === '||'))) {
+      result = '(' + result + ')';
+    }
```
Comment (rockwotj) [inline src/ast.ts]: Can we move this to the `else if (expOp.args.length === 2)` clause?

Also, this comment from above, seems like it's supposed to handle this:

```
// All ops are left associative - so nudge the innerPrecendence
// down on the right hand side to force () for right-associating
// operations.
```

@tomlarkworthy are you familiar with any of this code?

Comment (alixaxel) [reply src/ast.ts]: I noticed the comment and wondered about it, but as the tests show, it doesn't seem to be working as the comment describes.

I'll push the other change in a few hours.

**Comment by alixaxel** [comment]
@rockwotj I've updated the docs to describe the new output (some of it was already outdated, namely `Null` checks) and I've moved the condition to the the `expOp.args.length === 2` scope (I had originally done this, but ended up moving it further down to cover any unforeseen scenario).

**Comment by alixaxel** [comment]
@rockwotj Any update on this? :) I have a +6k LOC generated rules file that needs this fix to be deployed.

**Comment by rockwotj** [comment]
Thanks for the ping, this slipped. This has been released as `v0.8.4`.

https://github.com/firebase/bolt/releases/tag/v0.8.4

**Pull request merged**